### PR TITLE
revise copy for intro array challenge

### LINKED ISSUE #12629: Introduction missing for Basic Data Structures section (Beta site)
#### Section Name

Beta site Basic Data Structures

#### Issue Description

The Basic Data Structures section needs a landing page with introductory text to start the section. It should go in the slide format (challenge type 7) at the front of the section and give a high-level overview of what Basic Data Structures are.

Note: This section currently has the Debugging intro text, which should be removed.

### LINKED ISSUE #12693: [beta] final QA of basic DS section
The formatting in this section could use some work. Challenge instructions need key words wrapped in `<code>`, code examples should go in `<blockquote>`s like the rest of the challenges, and the test cases are not displaying correctly, they are displaying the asserts as well as the message (I think this is because `message: ....` was left out). This should be fixed in the seed code. As partially responsible for this section, I can fix this. 

image
Needs `<blockquote>` for example code ☝️ 

image
Test cases need to display correctly ☝️ 

image
Challenge instructions need `<code>` tags ☝️

### PR DESCRIPTION
**Pull request by no-stack-dub-sack**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Curriculum update
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Closes #12629 
- [x] Closes #12693 

#### Description

This is just a quick PR to update the copy for the Array section of the basic data structure challenges. I've been working through these for the last few days, and one thing I noticed is that the Object challenges gave a bit better of an overview on objects as data structures, so just wanted to match the tone and scope of that conversation (given the topic of the section) to make sure the array intro was just as thorough and complete.

### CODE REVIEW

**Comment by dhcodes** [comment]
I think at least part of both this challenge and the [Add-KeyValue-Pairs](https://beta.freecodecamp.com/en/challenges/challenges/basic-data-structures/add-keyvalue-pairs-to-javascript-objects) challenge should be converted into intro sections. There is **a lot** of text on these challenges and most of it doesn't actually have to do with the challenge exercise. Could this be split into a challenge and an intro like [this](https://beta.freecodecamp.com/en/challenges/object-oriented-programming/introduction-to-the-object-oriented-programming-challenges)

cc @freeCodeCamp/moderators Thoughts?

**Comment by raisedadead** [comment]
I agree with you @dhcodes , this is a bit wordy IMHO.

**Comment by no-stack-dub-sack** [comment]
@dhcodes @raisedadead I think that's an excellent idea and was def getting concerned with the wordiness of it all. If this is an option, I'd gladly do it. I think the information is important, and adds context to the section (otherwise they just become more object and array challenges). 

Only the most dedicated campers will probably end up reading it, but hey, at least no one can say we didn't provide the info

**Comment by no-stack-dub-sack** [comment]
@dhcodes @raisedadead Ok, take a look now. Modified to add an intro challenge to each section, making each first challenge's copy challenge more relevant to the challenge itself and, of course, less verbose.

Each one is a 3-click challenge with 2 short paragraphs on each page. Again, most will prob just skip, but I think it is better presentation.

**Comment by bonham000** [comment]
This looks good to me!

**Comment by HKuz** [comment]
@no-stack-dub-sack - with the intro done for this section, this PR can close out #12629, too 👍

**Comment by no-stack-dub-sack** [comment]
@HKuz awesome, thanks! Was actually going to go look and see if you had opened one for this section. Perfect!

**Comment by no-stack-dub-sack** [comment]
Wait... making a couple more changes. The second object challenge is also a bit long and really contains more "intro" material.

**Comment by no-stack-dub-sack** [comment]
@raisedadead Same question on this one - once #12720 is merged, will I need to rebase and force update this branch for the all of these changes to merge smoothly? Thanks! You've been a huge help to me lately

**Comment by raisedadead** [comment]
Yes, @no-stack-dub-sack and on that note may be the other PR #12720 can be taken up later. In which case we should merge this in and rebase that, which we are going to do in any case as it has conflicts.

Thanks and happy to help!

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 1]
```diff
@@ -5,17 +5,58 @@
...
+          ""
+        ],
+        [
+          "",
+          "",
+          "In Javascript, arrays are written as comma separated lists and are enclosed in brackets <code>[...]</code>. They can be any length, and store any type of data supported by JavaScript.<br><br>Arrays can sometimes be simple and serve very basic functionalities, but they can also be complex and very powerful - all of which depends on how the programmer chooses to utilize them<br><br>From this point forward, whenever we refer to the term <dfn>array</dfn>, it will be strictly within the context of JavaScript.",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: I think we should change this: `<code>[...]</code>` to either `<code>[a, b, c]</code>` or `<code>[element1, element2, element3]</code> to prevent confusion with the spread operator.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 2]
```diff
@@ -5,17 +5,58 @@
   "helpRoom": "Help",
   "challenges": [
     {
+      "id": "587d7dac367417b2b2516zx5",
+      "title": "Basic Data Structures: Introduction to Arrays",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: I think this can simply be "Introduction to Arrays" since the heading is above. (see: HTML section)

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 3]
```diff
@@ -394,18 +435,64 @@
       "translations": {}
     },
     {
+      "id": "587d7dac367417b2b25184d3",
+      "title": "Basic Data Structures: Introduction to Objects",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: Similar to above: I think this can be "Introduction to Objects". Remove "Basic Data Structures:"

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 4]
```diff
@@ -394,18 +435,64 @@
...
+      "description": [
+        [
+          "",
+          "",
+          "The next data structure we will discuss is the JavaScript <dfn>object</dfn>. Like arrays, objects are a very fundamental part of JavaScript. However, it is probably safe to say that objects surpass arrays in flexibility, usefuleness and in their overall importance to the language &mdash; in fact, you may have heard this line before: 'In JavaScript, everything is an object.'<br><br>While an understanding of objects is important to understand the inner workings of JavaScript functions or JavaScript's object-oriented capabilities, JavaScript objects at a basic level are actually just <dfn>key-value pair</dfn> stores, a commonly used data structure across almost all programming languages. Here, we will confine our discussion to JavaScript objects in this capacity.",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: Remove "very" in "objects are a very fundamental". Also, fix typo "usefuleness" should be "usefulness".

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 5]
```diff
@@ -394,18 +435,64 @@
...
+          ""
+        ],
+        [
+          "",
+          "",
+          "Another powerful advantage of key-value pair data structures is <dfn>constant lookup time</dfn>. What this means, is that when you request the value of a specific property, you will get the value back in the same amount of time (theoretically) regardless of the number of entries in the object. If you had an object with 5 entries or one that held a collection of 1,000,000, you could still retrieve property values or check if a key exists in the same amount of time.<br><br>The reason for this fast lookup time, is that internally, the object is storing properties using some type of hashing mechanism which allows it to know exactly where it has stored different property values. If you want to learn more about this please take a look at the optional Advanced Data Structures challenges. All you should remember for now is that <strong>performant access to flexibly structured data make key-value stores very attractive data structures useful in a wide variety of settings</strong>.",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: Change "the object is storing properties using some type of hashing mechanism..." to "the object is storing properties using a hashing mechanism...". Otherwise, it sounds unsure.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 6]
```diff
@@ -394,18 +435,64 @@
...
+          ""
+        ],
+        [
+          "",
+          "",
+          "<br><br>In JavaScript, objects are written as comma separated lists of key-value pairs, wrapped in curly brackets, with each key and its assigned value separated by a colon:<br> <code>{ key1: 'val-1', key2: 'val-2' }</code><br><br>In the next few challenges, we will examine JavaScript objects more closely, and take a look at a few key techniques and <dfn>methods</dfn> that allow us to access, store, and manipulate an object's data.<br><br>Note that throughout the scope of this discussion, and in general when considering JavaScript objects, he terms <dfn>key</dfn> and <dfn>property</dfn> will be used interchangably.",
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: "comma separated" should be "comma-separated". Also, change "at a few key techniques and" to "at"

Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: Also, fix typo "he" should be "the" in last line.

**Review by dhcodes — CHANGES_REQUESTED** [review]
Some wording fixes and other additions.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json (hunk 6 above)
Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: @dhcodes hahaha, I am beginning to hate this PR :smile: 

Thanks for the review though - solid points across the board. Will fix this in a few hours when I'm home from work and then _hopefully_ call it a day on this PR and this section!

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json (hunk 1 above)
Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: @dhcodes agreed, and meant to do this actually

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json (hunk 6 above)
Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: @dhcodes 
> Also, change "at a few key techniques and" to "at"
I disagree with this one though, because some of these challenges are not going over `methods` strictly speaking, but techniques instead

**Comment by no-stack-dub-sack** [comment]
@dhcodes ok, so I made all the changes you mentioned except for one - see last "outdated" comment above. So I think this should be good. And, you pointing out a few typos gave me a strong hunch that there would be more, so I spellchecked the whole thing and caught several more. If you and @raisedadead are ok with all of this, I think this section should be in pretty sound shape!

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
#### :sparkles: :sparkles: **LGTM.** :sparkles: :sparkles: 

Thanks @no-stack-dub-sack 

@dhcodes please have a look and merge.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json [hunk 7]
```diff
@@ -345,7 +386,7 @@
       ],
```
Comment (dhcodes) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: This challenge description has a mistake. 

References to `elem` should be changed to `num` or vice-versa.

Sorry for all the corrections @no-stack-dub-sack !

Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-data-structures.json]: @dhcodes haha, no worries, nice catch yet again, better to get this right! All set

**Comment by no-stack-dub-sack** [comment]
@dhcodes haha, no worries, nice catch yet again, better to get this right! All set

**Review by dhcodes — APPROVED** [review]

**Comment by dhcodes** [comment]
**LGTM** 👍 💯 🌮 Thanks for all the changes.
